### PR TITLE
plugin versions code reuse

### PR DESCRIPTION
**Pull request by brandonpollack23**
Get latest version if one not specified in pulumi manifest
Refactor code from

https://github.com/pulumi/pulumi/pull/16981 to share code.

Related to https://github.com/pulumi/pulumi/issues/16920

Original version: #16982

### CODE REVIEW

**Comment by brandonpollack23** [comment]
These failing tests do pass locally for me with the following command:

`PULUMI_GO_DEP_ROOT="/tmp/goroot" go test -tags=all -run ".*TestConfigMissingGo.*" github.com/pulumi/pulumi/tests/integration`

where /tmp/goroot contains a symlink to my pulumi checkout

File: pkg/engine/plugins.go [hunk 1]
```diff
@@ -215,6 +213,22 @@ func ensurePluginsAreInstalled(
 			continue
```
Comment (Frassle) [inline pkg/engine/plugins.go:213]: This continue means 'reinstall' is never hit.

File: pkg/engine/plugins.go [hunk 2]
```diff
@@ -190,15 +192,11 @@ func gatherPluginsFromSnapshot(plugctx *plugin.Context, target *deploy.Target) (
...
 // uses the given backend client to install them. Installations are processed in parallel, though
 // ensurePluginsAreInstalled does not return until all installations are completed.
-func ensurePluginsAreInstalled(
-	ctx context.Context,
-	opts *deploymentOptions,
-	d diag.Sink,
-	plugins pluginSet,
-	projectPlugins []workspace.ProjectPlugin,
+func EnsurePluginsAreInstalled(ctx context.Context, opts *deploymentOptions, d diag.Sink, plugins PluginSet,
+	projectPlugins []workspace.ProjectPlugin, reinstall, explicitInstall bool,
```
Comment (Frassle) [inline pkg/engine/plugins.go:199]:
```suggestion
	projectPlugins []workspace.ProjectPlugin, reinstall, installMissing bool,
```

Maybe a better name?

File: pkg/engine/update.go [hunk 3]
```diff
@@ -675,7 +677,8 @@ func (acts *updateActions) OnResourceStepPost(
 	// Write out the current snapshot. Note that even if a failure has occurred, we should still have a
 	// safe checkpoint.  Note that any error that occurs when writing the checkpoint trumps the error
 	// reported above.
-	return ctx.(SnapshotMutation).End(step, err == nil || status == resource.StatusPartialFailure)
+	return ctx.(SnapshotMutation).End(step, err == nil ||
```
Comment (Frassle) [inline pkg/engine/update.go:680]: Please don't make format changes unless needed, it makes blame more noisy.

File: pkg/engine/deployment.go [hunk 4]
```diff
@@ -55,7 +55,8 @@ func ProjectInfoContext(projinfo *Projinfo, host plugin.Host,
 
 	// Create a context for plugins.
 	ctx, err := plugin.NewContextWithRoot(diag, statusDiag, host, pwd, projinfo.Root,
-		projinfo.Proj.Runtime.Options(), disableProviderPreview, tracingSpan, projinfo.Proj.Plugins, config, debugging)
+		projinfo.Proj.Runtime.Options(), disableProviderPreview, tracingSpan,
```
Comment (Frassle) [inline pkg/engine/deployment.go:58]: Again no format changes unless needed.

File: pkg/engine/destroy.go [hunk 5]
```diff
@@ -80,8 +80,8 @@ func newDestroySource(
 
 	// Like Update, if we're missing plugins, attempt to download the missing plugins.
 
-	if err := ensurePluginsAreInstalled(ctx, opts, plugctx.Diag, plugins.Deduplicate(),
-		plugctx.Host.GetProjectPlugins()); err != nil {
+	if err := EnsurePluginsAreInstalled(ctx, opts, plugctx.Diag, plugins.Deduplicate(),
+		plugctx.Host.GetProjectPlugins(), false, false); err != nil {
```
Comment (Frassle) [inline pkg/engine/destroy.go:84]:
```suggestion
		plugctx.Host.GetProjectPlugins(), false /*reinstall*/, false /*explicitInstall*/); err != nil {
```

Also in the other places where we pass these as naked true/false

**Review by Frassle — COMMENTED** [review]
Couple of naked bool args to comment but otherwise I think this looks ok

**Comment by brandonpollack23** [comment]
fixed :)

**Review by Frassle — APPROVED** [review]

**Pull request merged**